Create rule S7038: "std::stringstream" or "std::spanstream" should be used instead of "std::strstream" (CPP-4998) (#4102)
SonarSource/rspec: rules/S7038/cfamily/rule.adoc

  os.freeze(false);
  return std::string(content);
}
----

==== Compliant solution

[source,cpp,diff-id=1,diff-type=compliant]
----
std::string printData(std::string_view entry, int count) {
  std::ostringstream os; // Compliant
  os << "The entry '" << entry << "' was repeated " << count << " times." << std::ends;
  return std::move(os).str();
}
----

=== Replacing output streams with preallocated buffer

In the case, when the `std::strstream` or `std::ostrstream` was used to put the string
into a preallocated buffer, exactly the same behavior may be preserved by using `std::spanstream` and `std::ospanstream`, respectively.

The constructor of ``++spanstream++``s accepts a `std::span<char>` instead of a pair of pointer and size.
Furthermore, the `str()` member function should be replaced with the `span()` member function, which returns a written range.

==== Noncompliant code example

[source,cpp,diff-id=2,diff-type=noncompliant]
----
std::string_view formatDateTo(int y, int m, int d, std::span<char> buffer) {
  std::ostrstream os(buffer.data(), buffer.size()); // Noncomplaint
  os << y << "/"  << m << "/" << d << std::ends;

  char const* data = os.str();

>   // Avoid buffer overflow, if the size of the buffer was not sufficient
>   std::size_t n = strnlen(data, buffer.size());

Reviewer: I wonder if we should explicitly mention this function is not standard... in an (indirect) way, this again strengthens the standard C++23 solution. Any opinion?

(C11's `strnlen_s` behaves differently)

```suggestion
  // Avoid buffer overflow, if the size of the buffer is not sufficient
  std::size_t n = strnlen(data, buffer.size());  // POSIX
```

Reviewer 2: Reworded this a bit to explain what `strnlen` does.

Reviewer: Looks good.

  return std::string_view(data, n);
}
----

==== Compliant solution

[source,cpp,diff-id=2,diff-type=compliant]
----
std::string_view formatDateTo(int y, int m, int d, std::span<char> buffer) {
  std::ospanstream os(buffer); // Compliant
  os << y << "/"  << m << "/" << d;
  return std::string_view(os.span());
}
----

We can avoid null-terminating the buffer (using `std::ends`), as `span()` provides both the pointer and the number of written characters.

==== Avoiding truncation of content

> Both in case of the `std::ostrstream` and `std::ospanstream` if the size of the buffer is insufficient for the content,
> the output will be trimmed. 
> If the use of a preallocated buffer is not strictly necessary due to performance reasons, you should consider using a `string stream`.

Reviewer: This is a really good point!

Minor grammar and stylistic suggestions:
```suggestion
With both `std::ostrstream` and `std::ospanstream`, the output is trimmed when the size of the buffer is insufficient for the content. 
If using a preallocated buffer is not strictly necessary for performance reasons, you should consider using a `stringstream`.
```

[source,cpp]
----
std::string formatDate(int y, int m, int d) {
  std::ostringstream os; // Compliant
  os << y << "/"  << m << "/" << d << std::ends;
  return std::move(os).str();
}
----

> === Replacing input streams with preallocated buffer

Reviewer: ```suggestion
=== Replacing input streams with preallocated buffers
```

or even this since it's always a preallocated buffer..
```suggestion
=== Replacing input streams
```

> The `std::ispanstream` should be used instead of `std::istrstream` to parse the content of the buffer.
> This requires converting the input arguments to the `std::span<const char>` which can be performed as follows:
>   * `std::span(ptr, n)` if `std::istrstream` was constructed from pointer `ptr` and `size`
>   * `std::string_view(cstr)` if `std::istrstream` was constructed from pointer `cstr`

Reviewer: ```suggestion
`std::ispanstream` should be used instead of `std::istrstream` to parse the content of a buffer.
This requires converting the input arguments to `std::span<const char>` which can be performed as follows:

  * `std::span(ptr, size)` if `std::istrstream` was constructed from pointer `ptr` and `size`;
  * `std::string_view(cstr)` if `std::istrstream` was constructed from null-terminated C-string `cstr`.
```

(The blank line is to fix the rendering.)

==== Noncompliant code example

[source,cpp,diff-id=3,diff-type=noncompliant]
----
int sum1(char const* buffer, int size) {
  int x, y;
  std::istrstream is(buffer, size); // Noncomplaint
  is >> x >> y;
  return x + y;
}

int sum2(char const* cstr) {
  int x, y;
  std::istrstream is(cstr); // Noncomplaint
  is >> x >> y;
  return x + y;
}
----

==== Compliant solution

[source,cpp,diff-id=3,diff-type=compliant]
----
int sum1(char const* buffer, int size) {
  int x, y;
  std::ispanstream is{std::span(buffer, size)}; // Compliant
  is >> x >> y;
  return x + y;
}

int sum2(char const* cstr) {
  int x, y;
  std::ispanstream is{std::string_view(cstr)}; // Compliant
  is >> x >> y;
  return x + y;
}
----

=== Fixing issue incrementally

> When the modified object is passed to the function, you may also need to adjust its signature.
> This, in turn, may require modifying other call sides of the given function and lead to large refactorings.
> In this section, we discuss a few options to mitigate the scale of the change.

Reviewer: ```suggestion
When the modified object is passed to a function, you may also need to adjust its signature.
This, in turn, may require modifying other call sites of the given function and lead to large refactorings.
In this section, we discuss a few options to mitigate the scale of the changes.
```

> === Passing reference to base class

Reviewer: Since it's meant to be a subsection, its heading should be h4

```suggestion
==== Passing reference to base class
```

> In a case when the body of the function does not use any functionality specific to `strstream` type,

Reviewer: ```suggestion
When the body of the function does not use any functionality specific to `strstream` type,
```

the best option may be to change the parameter to the reference to the corresponding base class:

[source,cpp,diff-id=4,diff-type=noncompliant]
----
void print(std::ostrstream& os, int n) {
  os << n;
  /* More streaming operations */
}
----

[source,cpp,diff-id=4,diff-type=compliant]
----
void print(std::ostream& os, int n) {
  os << n;
  /* More streaming operations */
}
----
 

> This change allows the above functions to be invoked with `stream`, `span stream`, and `string stream`.

Reviewer: ```suggestion
This change allows the above function to be invoked with `stream`, `spanstream`, and `stringstream`.
```

=== Introducing separate overloads

If the called function uses `strstream` specific functionality  (like calling `str()`),

> adding additional overload for corresponding `stringstream` or `spanstream` will allow incremental updates of the code.

Reviewer: ```suggestion
adding additional overload for corresponding `stringstream` or `spanstream` allows incremental improvements of the code.
```

To reduce code duplication, extract parts of the function that are not dependent on the `strstream` specific

> functionality into a helper function that accepts a corresponding base class reference.

Reviewer: ```suggestion
functionalities into a helper function that accepts a corresponding base class reference.
```

  int x, y;
  std::ispanstream is{std::span(buffer, size)}; // Compliant
  is >> x >> y;
  return x + y;
}

int sum2(char const* cstr) {
  int x, y;
  std::ispanstream is{std::string_view(cstr)}; // Compliant
  is >> x >> y;
  return x + y;
}
----

=== Fixing issue incrementally

When the modified object is passed to a function, you may also need to adjust its signature.
This, in turn, may require modifying other call sites of the given function and lead to large refactorings.
In this section, we discuss a few options to mitigate the scale of the changes.

==== Passing reference to base class

When the body of the function does not use any functionality specific to `strstream` type,
the best option may be to change the parameter to the reference to the corresponding base class:

[source,cpp,diff-id=4,diff-type=noncompliant]
----
void print(std::ostrstream& os, int n) {
  os << n;
  /* More streaming operations */
}
----

[source,cpp,diff-id=4,diff-type=compliant]
----
void print(std::ostream& os, int n) {
  os << n;
  /* More streaming operations */
}
----
 
This change allows the above function to be invoked with `stream`, `spanstream`, and `stringstream`.

> === Introducing separate overloads

Reviewer: This one should also be a subsection I think.
```suggestion
==== Introducing separate overloads
```

Uses of `std::strstream` and `std::ostrstream` should be replaced with:

 * `std::stringstream` and `std::ostringstream` to use dynamically allocated buffer
 * `std::spanstream` and `std::ospanstream` to use preallocated buffer.

Uses of `std::istrstream` should be replaced with `std::ispanstream`.

== Why is this an issue?

> The input `std::istrstream`, output `std::ostrstream` and combined `std::iostream` stream classes,
> are deprecated since they introduction in the first introduction in the {cpp}98 standard.

Reviewer: ```suggestion
The input `std::istrstream`, output `std::ostrstream` and combined `std::iostream` stream classes,
are deprecated since their introduction in the {cpp}98 standard.
```

or

```suggestion
The input `std::istrstream`, output `std::ostrstream` and combined `std::iostream` stream classes,
are deprecated since they were first introduced in the {cpp}98 standard.
```

(I tend to prefer the first version since it's shorter...)

> Above stream classes support two use cases, and their behavior differs vastly depending
> on the way how the object is constructed:

Reviewer: ```suggestion
Those stream classes support two use cases, and their behavior differs vastly depending
on how the object is constructed:
```

> * when constructed without argument, the stream object will manage dynamically allocated buffer
> * when constructed from `char*` buffer and size, the stream object will use only the provided buffer

Reviewer: ```suggestion
* When constructed without argument, the stream object manages a dynamically allocated buffer;
* When constructed from `char*` buffer and size, the stream object uses only the provided buffer.
```

> The above behavior made the `strstream` classes hard to use correctly.
> For example, invoking a `str()` method would either return a pointer to an internally managed dynamically allocated buffer,
> or a user-provided buffer. 

Reviewer: ```suggestion
The above behaviors make the `strstream` classes hard to use correctly.
For example, invoking a `str()` method can either return a pointer to an internally managed dynamically allocated buffer,
or a user-provided buffer. 
```

> or a user-provided buffer. 

Reviewer: I feel this doesn't really explain the problem: we should explain two things I think (from C++ reference):

> Before a call to str() that uses the result as a C-string, the stream buffer must be null-terminated. Regular output such as with stream << 1.2 does not store a null terminator, it must be appended explicitly, typically with the manipulator std::ends.
After a call to str(), dynamic streams become frozen. A call to freeze(false) is required before exiting the scope in which this strstream object was created. otherwise the destructor will leak memory. Also, additional output to a frozen stream may be truncated once it reaches the end of the allocated buffer, which may leave the buffer not null-terminated.

I am tempted to keep it really simple and simply quote C++ reference. Their example is also pretty convincing those classes should be avoided.

https://en.cppreference.com/w/cpp/io/strstream/str

WDYT?

Reviewer 2: This exists for `spanstream`, so it is not fixed.

Reviewer 2: Provided a better list of problems, that explicitly mentions that not giving size and guaranteeing null-termination is a problem.  Also added link to cppreference.

[source,cpp]
----
char* process(std::ostrstream& os) {
  os << "Important content";
  return os.str(); // Memory could be leaked, the ownership of returned pointer is not clear.
}
----

With the introduction of the `std::spanstream`, `std::ispanstream` and `std::ospanstream` {cpp}23,
all uses of the `std::strstream`, `std::ostrstream` and `std::ispanstream` classes may be replaced,
with one of the following classes:

* `std::stringstream` and `std::ostringstream` to use managed and growing buffer,
* `std::spanstream`, `std::ospanstream` and `std::ispanstream` to use preallocated buffer with fixed size.

> This rule will raise an issue when the new object of `strstream` is created with an underlying buffer:

Reviewer: I guess "dynamically-allocated" is better than "underlying" because a fixed-sized buffer is also "underlying".

However, I am not sure why you make the distinction. Shouldn't we raise on both kinds of creation?

Maybe that's what you meant... then I would drop the "with an underlying buffer" part, because there is no way to not have one.

Reviewer 2: Removed, I tried to differentiate creation by moving from another object.

[source,cpp]
----
std::string printData(std::string_view entry, int count) {
  std::ostrstream os; // Noncompliant
  os << "The entry '" << entry << "' was repeated " << count << " times." << std::ends;
  char const* content = os.str();
  os.freeze(false);
  return std::string(content);
}

int readInt(std::istrstream);

void process() {
  readInt(std::istrstream("10")); // Noncompliant
}t
----

Furthermore, the issue will be raised if a data member of the class is declared with one of `strstream` types:

[source,cpp]
----
class Printer {
  /* .... */
private:
  std::ostrstream os; // Noncomplaint
};
----

The rule does not raise an issue when `strsteam` object is moved or passed by reference, 

> as such use  will need to be adjusted after modifying the type of declared variable or created object.

Reviewer: I would drop this part. The rationale you gave me during our call makes sense, but this shorter version is not as clear and I fear it will only make the user wonder why.

----
int sum1(char const* buffer, int size) {
  int x, y;
  std::istrstream is(buffer, size); // Noncomplaint
  is >> x >> y;
  return x + y;
}

int sum2(char const* cstr) {
  int x, y;
  std::istrstream is(cstr); // Noncomplaint
  is >> x >> y;
  return x + y;
}
----

==== Compliant solution

[source,cpp,diff-id=3,diff-type=compliant]
----
int sum1(char const* buffer, int size) {
  int x, y;
  std::ispanstream is{std::span(buffer, size)}; // Compliant
  is >> x >> y;
  return x + y;
}

int sum2(char const* cstr) {
  int x, y;
  std::ispanstream is{std::string_view(cstr)}; // Compliant
  is >> x >> y;
  return x + y;
}
----

=== Fixing issue incrementally

When the modified object is passed to the function, you may also need to adjust the it's signature.
This in turn may require modifying other call sides of the given function, and lead to large refactorings.
In this section we disucss few options to mitigate the scale of the change.

=== Passing reference to base class

In a case when the body of the function does not use any funcitonality specific to `strstream` type,
the best option may be to change the parameter to the reference to the corresponding base class:

> <<<<<<< HEAD

Reviewer: There are some unresolved merged conflicts.

Reviewer 2: Fixed the conflict. This part can be reviewed.

== Why is this an issue?

The input `std::istrstream`, output `std::ostrstream` and combined `std::iostream` stream classes,
are deprecated since their introduction in the {cpp}98 standard.

Those stream classes support two use cases, and their behavior differs vastly depending
on how the object is constructed:

* When constructed without argument, the stream object manages a dynamically allocated buffer;
* When constructed from ``++char*++`` buffer and size, the stream object uses only the provided buffer.

The above behaviors make the `strstream` classes hard to use correctly.
As illustration `str()` member function:

* returns a buffer that has unclear ownership
* does not return the size of readable characters nor guarantees null-termination of buffer

> * may leak unless `freeze(false)` is called afterwards

Reviewer 2: Applied manually

}
----
 
This change allows the above function to be invoked with `stream`, `spanstream`, and `stringstream`.

=== Introducing separate overloads

If the called function uses `strstream` specific functionality  (like calling `str()`),
adding additional overload for corresponding `stringstream` or `spanstream` allows incremental improvements of the code.

To reduce code duplication, extract parts of the function that are not dependent on the `strstream` specific

> functionalities into a helper function that accepts a referenc to corresponding base class.

Reviewer: ```suggestion
functionalities into a helper function that accepts a reference to the corresponding base class.
```
